File: backend/app/schemas/gap_reconstruction.py
from __future__ import annotations

from datetime import datetime
from enum import StrEnum
from math import isfinite
from typing import Any

from pydantic import Field, field_validator, model_validator

from .common import ApiModel


class Availability(StrEnum):
    AVAILABLE = "AVAILABLE"
    UNAVAILABLE = "UNAVAILABLE"
    UNKNOWN = "UNKNOWN"


class ReviewStatus(StrEnum):
    NOT_REVIEWED = "NOT_REVIEWED"
    REQUESTED = "REQUESTED"
    FOOTAGE_UNAVAILABLE = "FOOTAGE_UNAVAILABLE"
    REVIEWED_NO_RELEVANT_FINDING = "REVIEWED_NO_RELEVANT_FINDING"
    RELEVANT_FOOTAGE_FOUND = "RELEVANT_FOOTAGE_FOUND"


class Coordinates(ApiModel):
    latitude: float
    longitude: float

    @model_validator(mode="after")
    def valid(self) -> "Coordinates":
        if not isfinite(self.latitude) or not isfinite(self.longitude) or not -90 <= self.latitude <= 90 or not -180 <= self.longitude <= 180:
            raise ValueError("Coordinates must be finite and valid")
        return self


class CameraCreate(ApiModel):
    label: str = Field(min_length=1, max_length=160)
    coordinates: Coordinates
    source_ref: str = Field(min_length=1, max_length=500)
    is_synthetic: bool = False
    operational_from: datetime | None = None
    operational_to: datetime | None = None
    recording_availability: Availability = Availability.UNKNOWN
    retention_information: str | None = Field(default=None, max_length=1000)
    verified_orientation: dict[str, Any] | None = None

    @field_validator("operational_from", "operational_to")
    @classmethod
    def operational_dates_are_aware(cls, value: datetime | None) -> datetime | None:
        if value is not None and value.tzinfo is None:
            raise ValueError("Operational dates must include timezone offsets")
        return value

    @model_validator(mode="after")
    def dates_in_order(self) -> "CameraCreate":
        if self.operational_from and self.operational_to and self.operational_to < self.operational_from:
            raise ValueError("Operational end must not precede start")
        return self


class CameraUpdate(ApiModel):
    label: str | None = Field(default=None, min_length=1, max_length=160)
    coordinates: Coordinates | None = None
    source_ref: str | None = Field(default=None, min_length=1, max_length=500)
    is_synthetic: bool | None = None
    recording_availability: Availability | None = None
    retention_information: str | None = Field(default=None, max_length=1000)
    operational_from: datetime | None = None
    operational_to: datetime | None = None
    verified_orientation: dict[str, Any] | None = None

    @field_validator("operational_from", "operational_to")
    @classmethod
    def operational_dates_are_aware(cls, value: datetime | None) -> datetime | None:
        if value is not None and value.tzinfo is None:
            raise ValueError("Operational dates must include timezone offsets")
        return value


class Camera(CameraCreate):
    id: str
    investigation_id: str
    created_at: datetime
    updated_at: datetime


class ObservationInput(ApiModel):
    event_id: str = Field(min_length=1, max_length=160)
    location_id: str = Field(min_length=1, max_length=160)
    observed_at: datetime
    earliest_at: datetime | None = None
    latest_at: datetime | None = None
    time_precision: str
    association_kind: str = Field(pattern="^(DOCUMENTED_SIGHTING|ANALYST_ASSUMPTION)$")
    analyst_assumption_note: str | None = Field(default=None, max_length=2000)

    @model_validator(mode="after")
    def bounded_time(self) -> "ObservationInput":
        if self.observed_at.tzinfo is None:
            raise ValueError("Observation timestamps must include timezone offsets")
        if self.time_precision != "EXACT":
            if self.earliest_at is None or self.latest_at is None:
                raise ValueError("Approximate observations require explicit earliestAt and latestAt bounds")
            if self.earliest_at.tzinfo is None or self.latest_at.tzinfo is None or self.earliest_at > self.latest_at:
                raise ValueError("Approximate time bounds must be timezone-aware and ordered")
            if not self.earliest_at <= self.observed_at <= self.latest_at:
                raise ValueError("Observed time must fall within the supplied bounds")
        if self.association_kind == "ANALYST_ASSUMPTION" and not (self.analyst_assumption_note or "").strip():
            raise ValueError("Analyst assumptions require a note")
        return self


class ReconstructionRequest(ApiModel):
    entity_id: str = Field(min_length=1, max_length=160)
    start_observation: ObservationInput
    end_observation: ObservationInput
    travel_mode: str = Field(min_length=1, max_length=80)
    speed_kmh: float = Field(gt=0, le=400)
    selected_time: datetime | None = None


class CandidateResult(ApiModel):
    camera: Camera
    included: bool
    reason: str
    distance_from_start_km: float
    distance_to_end_km: float
    earliest_arrival: datetime | None = None
    latest_departure: datetime | None = None
    operational_status: str
    recording_availability: Availability


class ReconstructionPreview(ApiModel):
    investigation_id: str
    algorithm_version: str
    model_label: str
    is_envelope: bool
    feasible: bool
    reason: str
    start_coordinates: Coordinates
    end_coordinates: Coordinates
    start_time: datetime
    end_time: datetime
    selected_time: datetime
    forward_radius_km: float
    backward_radius_km: float
    forward_region: list[list[float]]
    backward_region: list[list[float]]
    candidates: list[CandidateResult]
    excluded_cameras: list[CandidateResult]
    inputs: ReconstructionRequest


class SavedReconstruction(ReconstructionPreview):
    id: str
    investigator_user_id: str
    created_at: datetime
    reviews: list[dict[str, Any]] = Field(default_factory=list)


class ReviewUpdate(ApiModel):
    status: ReviewStatus
    notes: str = Field(default="", max_length=4000)
    evidence_id: str | None = None


class CameraReview(ApiModel):
    run_id: str
    camera_id: str
    investigation_id: str
    status: ReviewStatus
    notes: str
    evidence_id: str | None
    reviewed_by: str
    updated_at: datetime

File: lib/gap-reconstruction.ts
import type { InvestigationId } from "@/data/investigations/types";

export type CameraAvailability = "AVAILABLE" | "UNAVAILABLE" | "UNKNOWN";
export type CameraReviewStatus = "NOT_REVIEWED" | "REQUESTED" | "FOOTAGE_UNAVAILABLE" | "REVIEWED_NO_RELEVANT_FINDING" | "RELEVANT_FOOTAGE_FOUND";
export type GapCoordinates = { latitude: number; longitude: number };
export type GapCamera = { id: string; investigationId: InvestigationId; label: string; coordinates: GapCoordinates; sourceRef: string; isSynthetic: boolean; operationalFrom: string | null; operationalTo: string | null; recordingAvailability: CameraAvailability; retentionInformation: string | null; verifiedOrientation: Record<string, unknown> | null; createdAt: string; updatedAt: string };
export type GapObservation = { eventId: string; locationId: string; observedAt: string; earliestAt?: string; latestAt?: string; timePrecision: string; associationKind: "DOCUMENTED_SIGHTING" | "ANALYST_ASSUMPTION"; analystAssumptionNote?: string };
export type GapRequest = { entityId: string; startObservation: GapObservation; endObservation: GapObservation; travelMode: string; speedKmh: number; selectedTime?: string };
export type GapCandidate = { camera: GapCamera; included: boolean; reason: string; distanceFromStartKm: number; distanceToEndKm: number; earliestArrival: string | null; latestDeparture: string | null; operationalStatus: string; recordingAvailability: CameraAvailability };
export type GapPreview = { investigationId: InvestigationId; algorithmVersion: string; modelLabel: string; isEnvelope: boolean; feasible: boolean; reason: string; startCoordinates: GapCoordinates; endCoordinates: GapCoordinates; startTime: string; endTime: string; selectedTime: string; forwardRadiusKm: number; backwardRadiusKm: number; forwardRegion: number[][]; backwardRegion: number[][]; candidates: GapCandidate[]; excludedCameras: GapCandidate[]; inputs: GapRequest; id?: string; investigatorUserId?: string; createdAt?: string; reviews?: GapCameraReview[] };
export type GapCameraReview = { runId: string; cameraId: string; investigationId: string; status: CameraReviewStatus; notes: string; evidenceId: string | null; reviewedBy: string; updatedAt: string };

const EARTH_RADIUS_KM = 6371.0088;

function assertPoint(point: GapCoordinates) {
  if (!Number.isFinite(point.latitude) || !Number.isFinite(point.longitude) || Math.abs(point.latitude) > 90 || Math.abs(point.longitude) > 180) throw new Error("Invalid coordinates");
}

export function haversineDistanceKm(a: GapCoordinates, b: GapCoordinates) {
  assertPoint(a); assertPoint(b);
  const rad = Math.PI / 180;
  const lat1 = a.latitude * rad; const lat2 = b.latitude * rad;
  const dLat = (b.latitude - a.latitude) * rad; const dLon = (b.longitude - a.longitude) * rad;
  const value = Math.sin(dLat / 2) ** 2 + Math.cos(lat1) * Math.cos(lat2) * Math.sin(dLon / 2) ** 2;
  return EARTH_RADIUS_KM * 2 * Math.asin(Math.sqrt(Math.min(1, Math.max(0, value))));
}

export function geodesicCircle(center: GapCoordinates, radiusKm: number, vertices = 96): [number, number][] {
  assertPoint(center);
  if (!Number.isFinite(radiusKm) || radiusKm < 0) throw new Error("Invalid radius");
  const lat1 = center.latitude * Math.PI / 180; const lon1 = center.longitude * Math.PI / 180;
  const angular = radiusKm / EARTH_RADIUS_KM;
  return Array.from({ length: vertices + 1 }, (_, index) => {
    const bearing = index * 2 * Math.PI / vertices;
    const lat2 = Math.asin(Math.sin(lat1) * Math.cos(angular) + Math.cos(lat1) * Math.sin(angular) * Math.cos(bearing));
    const lon2 = lon1 + Math.atan2(Math.sin(bearing) * Math.sin(angular) * Math.cos(lat1), Math.cos(angular) - Math.sin(lat1) * Math.sin(lat2));
    return [((lon2 * 180 / Math.PI + 540) % 360) - 180, lat2 * 180 / Math.PI];
  });
}

export function reachabilityAt(preview: GapPreview, selectedTime: string) {
  const start = Date.parse(preview.startTime); const end = Date.parse(preview.endTime); const at = Date.parse(selectedTime);
  if (![start, end, at].every(Number.isFinite) || at < start || at > end || end <= start) throw new Error("Invalid analysis time");
  const speed = preview.inputs.speedKmh;
  const forwardRadiusKm = speed * (at - start) / 3_600_000;
  const backwardRadiusKm = speed * (end - at) / 3_600_000;
  return { forwardRadiusKm, backwardRadiusKm, forwardRegion: geodesicCircle(preview.startCoordinates, forwardRadiusKm), backwardRegion: geodesicCircle(preview.endCoordinates, backwardRadiusKm) };
}
